lighthouse-core/gather/computed/first-interactive.js

```diff
@@ -0,0 +1,222 @@
+        const start = tasks[0].start;
+        const end = tasks[tasks.length - 1].end;
+        const duration = end - start;
+        return {start, end, duration};
+      })
+      // filter out clusters that started after the window because of our clusteringWindowEnd
+      .filter(cluster => cluster.start < windowEnd);
+  }
+
+  /**
+   * Finds the timeInMs of the start of the first quiet window as defined by the firstInteractive
+   * conditions above. Throws an error if no acceptable quiet window could be found before the end
+   * of the trace.
+   * @param {number} FMP
+   * @param {number} traceEnd
+   * @param {!Array<{start: number, end: number>}} longTasks
+   * @return {number}
+   */
+  static findQuietWindow(FMP, traceEnd, longTasks) {
+    // If we have an empty window at the very beginning, just return FMP early
+    if (longTasks.length === 0 ||
+        longTasks[0].start > FMP + FirstInteractive.getRequiredWindowSizeInMs(0)) {
+      return FMP;
+    }
+
+    const isTooCloseToFMP = cluster => cluster.start < FMP + MIN_TASK_CLUSTER_FMP_DISTANCE;
+    const isTooLong = cluster => cluster.duration > MAX_TASK_CLUSTER_DURATION;
+    const isBadCluster = cluster => isTooCloseToFMP(cluster) || isTooLong(cluster);
+
+    // FirstInteractive must start at the end of a long task, consider each long task and
+    // examine the window that follows it.
+    for (let i = 0; i < longTasks.length; i++) {
+      const windowStart = longTasks[i].end;
+      const windowSize = FirstInteractive.getRequiredWindowSizeInMs(windowStart - FMP);
+      const windowEnd = windowStart + windowSize;
+
+      // Check that we have a long enough trace
+      if (windowEnd > traceEnd) {
+        throw new Error(TRACE_BUSY_MSG);
+      }
+
+      // Check that this task isn't the beginning of a cluster
+      if (i + 1 < longTasks.length &&
+          longTasks[i + 1].start - windowStart <= MIN_TASK_CLUSTER_PADDING) {
+        continue;
+      }
+
+      const taskClusters = FirstInteractive.getTaskClustersInWindow(longTasks, i + 1, windowEnd);
+      const hasBadTaskClusters = taskClusters.some(isBadCluster);
+
+      if (!hasBadTaskClusters) {
+        return windowStart;
+      }
+    }
+
+    throw new Error(TRACE_BUSY_MSG);
+  }
+
+  /**
+   * @param {!Trace} trace
+   * @param {!tr.Model} traceModel
+   * @param {!TraceOfTabArtifact} traceOfTab
+   * @return {{timeInMs: number, timestamp: number}}
+   */
+  computeWithArtifacts(trace, traceModel, traceOfTab) {
+    const navStart = traceOfTab.timestamps.navigationStart;
+    const FMP = traceOfTab.timings.firstMeaningfulPaint;
+    const DCL = traceOfTab.timings.domContentLoaded;
+    const traceEnd = traceOfTab.timings.traceEnd;
+
+    if (traceEnd - FMP < MAX_QUIET_WINDOW_SIZE) {
+      throw new Error('trace not at least 5 seconds longer than FMP');
+    }
+
+    if (!FMP || !DCL) {
+      throw new Error(`No ${FMP ? 'domContentLoaded' : 'firstMeaningfulPaint'} event in trace`);
```

Comment:
I only moved FMP since DCL is specific to TTFI but everything needs FMP basically